File: packages/sdk/ruby/packages/core/lib/edgebase_core/http_client.rb
# frozen_string_literal: true

require "net/http"
require "uri"
require "json"

require_relative "context_manager"
require_relative "errors"

module EdgebaseCore
  DEFAULT_OPEN_TIMEOUT = 30
  DEFAULT_READ_TIMEOUT = 120

  # Synchronous HTTP client for server-side use.
  #
  # Features:
  # - Service Key header injection (X-EdgeBase-Service-Key)
  # - Optional Bearer token injection (for impersonation)
  # - Legacy context state for compatibility (not serialized into HTTP headers)
  class HttpClient
    attr_reader :base_url

    def initialize(base_url, context_manager: nil, service_key: nil, bearer_token: nil)
      @base_url = base_url.chomp("/")
      @context_manager = context_manager || ContextManager.new
      @service_key = service_key
      @bearer_token = bearer_token
    end

    def get(path, params: nil)
      request("GET", path, params: params)
    end

    def post(path, body = nil, params: nil)
      request("POST", path, params: params, json_body: body)
    end

    def patch(path, body = nil, params: nil)
      request("PATCH", path, params: params, json_body: body)
    end

    def put(path, body = nil, params: nil)
      request("PUT", path, params: params, json_body: body)
    end

    def delete(path, params: nil)
      request("DELETE", path, params: params)
    end

    # HEAD request — returns true if resource exists (2xx).
    def head(path)
      uri = URI(build_url(path))
      http = build_http(uri)
      req = Net::HTTP::Head.new(uri)
      auth_headers.each { |k, v| req[k] = v }
      response = http.request(req)
      response.code.to_i < 400
    end

    # POST multipart form data (for file uploads).
    def post_multipart(path, files:, data: nil)
      uri = URI(build_url(path))
      http = build_http(uri)

      boundary = "EdgeBase#{rand(10**16)}"
      body = build_multipart_body(files, data, boundary)

      req = Net::HTTP::Post.new(uri)
      headers = auth_headers
      headers.delete("Content-Type")
      headers["Content-Type"] = "multipart/form-data; boundary=#{boundary}"
      headers.each { |k, v| req[k] = v }
      req.body = body

      parse_response(http.request(req))
    end

    # POST raw binary data (for multipart upload-part).
    def post_raw(path, data:, content_type: "application/octet-stream")
      uri = URI(build_url(path))
      http = build_http(uri)

      req = Net::HTTP::Post.new(uri)
      headers = auth_headers
      headers["Content-Type"] = content_type
      headers.each { |k, v| req[k] = v }
      req.body = data

      parse_response(http.request(req))
    end

    # GET raw bytes (for file downloads).
    def get_raw(path)
      uri = URI(build_url(path))
      http = build_http(uri)
      req = Net::HTTP::Get.new(uri)
      auth_headers.each { |k, v| req[k] = v }
      response = http.request(req)
      if response.code.to_i >= 400
        raise EdgeBaseError.new(response.code.to_i, response.body)
      end
      response.body
    end

    private

    def parse_retry_after_delay(response, attempt)
      base_delay = 1.0 * (2 ** attempt)
      retry_after = response["retry-after"] || response["Retry-After"]
      if retry_after
        begin
          seconds = Integer(retry_after)
          base_delay = seconds.to_f if seconds > 0
        rescue ArgumentError
          # ignore non-integer Retry-After
        end
      end
      jitter = rand * base_delay * 0.25
      [base_delay + jitter, 10.0].min
    end

    RETRYABLE_ERRORS = [
      Net::OpenTimeout, Net::ReadTimeout, Errno::ECONNREFUSED,
      Errno::ECONNRESET, Errno::ENETUNREACH, Errno::EHOSTUNREACH,
      SocketError, EOFError
    ].freeze

    def request(method, path, params: nil, json_body: nil)
      url = build_url(path)
      if params && !params.empty?
        query = URI.encode_www_form(params)
        url = "#{url}?#{query}"
      end

      max_retries = 3
      last_response = nil

      (max_retries + 1).times do |attempt|
        uri = URI(url)
        http = build_http(uri)

        req = case method
              when "GET"    then Net::HTTP::Get.new(uri)
              when "POST"   then Net::HTTP::Post.new(uri)
              when "PATCH"  then Net::HTTP::Patch.new(uri)
              when "PUT"    then Net::HTTP::Put.new(uri)
              when "DELETE" then Net::HTTP::Delete.new(uri)
              else Net::HTTP::Get.new(uri)
              end

        auth_headers.each { |k, v| req[k] = v }
        req.body = JSON.generate(json_body) if json_body

        begin
          response = http.request(req)
          last_response = response
        rescue *RETRYABLE_ERRORS => e
          if attempt < 2
            sleep(0.05 * (attempt + 1))
            next
          end
          raise EdgeBaseError.new(0, "Request failed: #{e.message}")
        end

        if response.code.to_i == 429 && attempt < max_retries
          delay = parse_retry_after_delay(response, attempt)
          sleep(delay)
          next
        end

        return parse_response(response)
      end

      parse_response(last_response)
    end

    def build_url(path)
      if path.start_with?("/api/")
        "#{@base_url}#{path}"
      else
        "#{@base_url}/api#{path}"
      end
    end

    def auth_headers
      headers = { "Content-Type" => "application/json", "Connection" => "close" }
      begin
        if @bearer_token
          headers["Authorization"] = "Bearer #{@bearer_token}"
        end
        if @service_key
          headers["X-EdgeBase-Service-Key"] = @service_key
          headers["Authorization"] = "Bearer #{@service_key}"
        end
      rescue StandardError
        # Token refresh failed — proceed as unauthenticated
      end
      headers
    end

    def build_http(uri)
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = (uri.scheme == "https")
      http.open_timeout = request_timeout_seconds(DEFAULT_OPEN_TIMEOUT)
      http.read_timeout = request_timeout_seconds(DEFAULT_READ_TIMEOUT)
      http.keep_alive_timeout = 0 if http.respond_to?(:keep_alive_timeout=)
      http
    end

    def request_timeout_seconds(default_seconds)
      raw = ENV.fetch("EDGEBASE_HTTP_TIMEOUT_MS", "").strip
      return default_seconds if raw.empty?

      timeout_ms = Integer(raw, exception: false)
      return default_seconds if timeout_ms.nil? || timeout_ms <= 0

      timeout_ms / 1000.0
    end

    def parse_response(response)
      code = response.code.to_i
      if code >= 400
        begin
          data = JSON.parse(response.body)
          raise EdgeBaseError.from_json(data, code)
        rescue EdgeBaseError
          raise
        rescue StandardError
          raise EdgeBaseError.new(code, response.body || "Request failed with HTTP #{code} and a non-JSON error response.")
        end
      end

      return nil if response.body.nil? || response.body.empty?

      begin
        JSON.parse(response.body)
      rescue StandardError
        raise EdgeBaseError.new(code, "Expected a JSON response but received malformed JSON.")
      end
    end

    def build_multipart_body(files, data, boundary)
      body = +""
      # Data fields
      if data
        data.each do |key, value|
          body << "--#{boundary}\r\n"
          body << "Content-Disposition: form-data; name=\"#{key}\"\r\n\r\n"
          body << "#{value}\r\n"
        end
      end
      # File fields
      files.each do |field_name, (filename, file_data, content_type)|
        body << "--#{boundary}\r\n"
        body << "Content-Disposition: form-data; name=\"#{field_name}\"; filename=\"#{filename}\"\r\n"
        body << "Content-Type: #{content_type}\r\n\r\n"
        body << file_data
        body << "\r\n"
      end
      body << "--#{boundary}--\r\n"
      body
    end
  end
end

File: packages/sdk/python/packages/core/src/edgebase_core/http_client.py
"""HTTP client — httpx-based, server-side only.

Python SDK is server-only: authenticates via Service Key header.
No token refresh logic — use service_key for all server-side calls.
"""

from __future__ import annotations

import os
import time
import random
from typing import Any

import httpx

from edgebase_core.context_manager import ContextManager
from edgebase_core.errors import EdgeBaseError

DEFAULT_HTTP_TIMEOUT = httpx.Timeout(120.0, connect=30.0)
DEFAULT_HTTP_LIMITS = httpx.Limits(max_connections=10, max_keepalive_connections=0, keepalive_expiry=0.0)


class HttpClient:
    """Synchronous HTTP client for server-side use.

    Features:
    - Service Key header injection (X-EdgeBase-Service-Key)
    - Optional Bearer token injection (for impersonation)
    - Legacy context state for compatibility (not serialized into HTTP headers)
    """

    def __init__(
        self,
        base_url: str,
        context_manager: ContextManager | None = None,
        service_key: str | None = None,
        bearer_token: str | None = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._context_manager = context_manager if context_manager is not None else ContextManager()
        self._service_key = service_key
        self._bearer_token = bearer_token
        self._client = httpx.Client(
            timeout=self._resolve_timeout(),
            limits=DEFAULT_HTTP_LIMITS,
            headers={"Connection": "close"},
        )

    @staticmethod
    def _resolve_timeout() -> httpx.Timeout:
        raw = os.getenv("EDGEBASE_HTTP_TIMEOUT_MS", "").strip()
        if not raw:
            return DEFAULT_HTTP_TIMEOUT
        try:
            timeout_ms = int(raw)
        except ValueError:
            return DEFAULT_HTTP_TIMEOUT
        if timeout_ms <= 0:
            return DEFAULT_HTTP_TIMEOUT
        return httpx.Timeout(timeout_ms / 1000.0)

    def get(self, path: str, params: dict[str, str] | None = None) -> Any:
        return self._request("GET", path, params=params)

    def post(self, path: str, body: Any = None, params: dict[str, str] | None = None) -> Any:
        return self._request("POST", path, json_body=body, params=params)

    def patch(self, path: str, body: Any = None, params: dict[str, str] | None = None) -> Any:
        return self._request("PATCH", path, json_body=body, params=params)

    def put(self, path: str, body: Any = None, params: dict[str, str] | None = None) -> Any:
        return self._request("PUT", path, json_body=body, params=params)

    def delete(self, path: str, body: Any = None) -> Any:
        return self._request("DELETE", path, json_body=body)

    def head(self, path: str) -> bool:
        """HEAD request — returns True if resource exists (2xx)."""
        url = self._build_url(path)
        headers = self._auth_headers()
        response = self._client.request("HEAD", url, headers=headers)
        return response.status_code < 400

    def post_multipart(
        self,
        path: str,
        files: dict[str, Any],
        data: dict[str, str] | None = None,
    ) -> Any:
        """POST multipart form data (for file uploads)."""
        url = self._build_url(path)
        headers = self._auth_headers()
        headers.pop("Content-Type", None)  # Let httpx set multipart boundary
        response = self._client.post(url, files=files, data=data, headers=headers)
        return self._parse_response(response)

    def post_raw(
        self,
        path: str,
        data: bytes,
        content_type: str = "application/octet-stream",
    ) -> Any:
        """POST raw binary data (for multipart upload-part)."""
        url = self._build_url(path)
        headers = self._auth_headers()
        headers["Content-Type"] = content_type
        response = self._client.post(url, content=data, headers=headers)
        return self._parse_response(response)

    def get_raw(self, path: str) -> bytes:
        """GET raw bytes (for file downloads)."""
        url = self._build_url(path)
        headers = self._auth_headers()
        response = self._client.get(url, headers=headers)
        if response.status_code >= 400:
            raise EdgeBaseError(response.status_code, response.text)
        return response.content

    def close(self) -> None:
        """Close the underlying httpx client."""
        self._client.close()

    # MARK: - Internal

    @staticmethod
    def _parse_retry_after_delay(response: httpx.Response, attempt: int) -> float:
        base_delay = 1.0 * (2 ** attempt)
        retry_after = response.headers.get("retry-after")
        if retry_after:
            try:
                seconds = int(retry_after)
                if seconds > 0: base_delay = float(seconds)
            except ValueError: pass
        jitter = random.random() * base_delay * 0.25
        return min(base_delay + jitter, 10.0)

    @staticmethod
    def _is_retryable_transport_error(error: Exception) -> bool:
        msg = str(error).lower()
        return any(k in msg for k in ["timeout", "connection", "reset", "refused", "network", "eof"])

    def _request(self, method: str, path: str, params: dict[str, str] | None = None, json_body: Any = None) -> Any:
        url = self._build_url(path)
        headers = self._auth_headers()
        max_retries = 3
        for attempt in range(max_retries + 1):
            try:
                response = self._client.request(method, url, params=params, json=json_body, headers=headers)
            except Exception as exc:
                if attempt < 2 and self._is_retryable_transport_error(exc):
                    time.sleep(0.05 * (attempt + 1))
                    continue
                raise
            if response.status_code == 429 and attempt < max_retries:
                delay = self._parse_retry_after_delay(response, attempt)
                time.sleep(delay)
                continue
            return self._parse_response(response)
        # Final attempt exhausted (e.g., persistent 429) — parse last response as-is
        return self._parse_response(response)

    def _build_url(self, path: str) -> str:
        if path.startswith("/api/"):
            return f"{self._base_url}{path}"
        return f"{self._base_url}/api{path}"

    def _auth_headers(self) -> dict[str, str]:
        headers: dict[str, str] = {
            "Content-Type": "application/json",
            "Connection": "close",
        }
        try:
            if self._bearer_token:
                headers["Authorization"] = f"Bearer {self._bearer_token}"
            if self._service_key:
                headers["X-EdgeBase-Service-Key"] = self._service_key
                headers["Authorization"] = f"Bearer {self._service_key}"
        except Exception:
            # Token refresh failed — proceed as unauthenticated
            pass
        return headers

    @staticmethod
    def _parse_response(response: httpx.Response) -> Any:
        if response.status_code >= 400:
            try:
                data = response.json()
                raise EdgeBaseError.from_json(data, response.status_code)
            except EdgeBaseError:
                raise
            except Exception:
                raise EdgeBaseError(response.status_code, response.text)

        if not response.content or response.status_code == 204:
            return None
        try:
            return response.json()
        except Exception:
            raise EdgeBaseError(
                response.status_code,
                "Expected a JSON response but received malformed JSON.",
            )
